### PR TITLE
refactor: error/anyhow::Error wrapping

### PR DESCRIPTION
**Pull request by koivunej**
We can get CompactionError::Other(Cancelled) via the error handling with a few ways. [evidence](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8655/10301613380/index.html#suites/cae012a1e6acdd9fdd8b81541972b6ce/653a33de17802bb1/). Hopefully fix it by:

1. replace the `map_err` which hid the `GetReadyAncestorError::Cancelled` with `From<GetReadyAncestorError> for GetVectoredError` conversion
2. simplifying the code in pgdatadir_mapping to eliminate the token anyhow wrapping for deserialization errors
3. stop wrapping GetVectoredError as anyhow errors
4. stop wrapping PageReconstructError as anyhow errors

Additionally, produce warnings if we treat any other error (as was legal before this PR) as missing key.

Cc: #8708.

### CODE REVIEW

**Comment by koivunej** [comment]
> * `test_branch_and_gc`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8697/10352193815/index.html#suites/cae012a1e6acdd9fdd8b81541972b6ce/7ae28c662155b354)

Again...

**Comment by koivunej** [comment]
Having trouble reproducing that message.

**Comment by koivunej** [comment]
This is quite elusive. Reproduced locally. From the latest commit:

```
2024-08-13T07:41:08.033600Z ERROR compaction_loop{tenant_id=c547a720666f14ca442ac2c4a6c44d7c shard_id=0000}: Compaction failed 1 times, retrying in 2s: Other(Cancelled

Stack backtrace:
   0: <T as core::convert::Into<U>>::into
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/convert/mod.rs:759:9
      <pageserver::tenant::timeline::CompactionError as core::convert::From<pageserver::tenant::timeline::CreateImageLayersError>>::from
             at /home/joonas/src/neon/work/pageserver/src/tenant/timeline.rs:763:43
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
             at /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/result.rs:1989:27
      pageserver::tenant::timeline::compaction::<impl pageserver::tenant::timeline::Timeline>::compact_legacy::{{closure}}
             at /home/joonas/src/neon/work/pageserver/src/tenant/timeline/compaction.rs:298:40
      pageserver::tenant::timeline::Timeline::compact::{{closure}}
             at /home/joonas/src/neon/work/pageserver/src/tenant/timeline.rs:1583:84
   2: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
             at /home/joonas/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tracing-0.1.37/src/instrument.rs:272:9
      pageserver::tenant::Tenant::compaction_iteration::{{closure}}
             at /home/joonas/src/neon/work/pageserver/src/tenant.rs:1788:18
      pageserver::tenant::tasks::compaction_loop::{{closure}}::{{closure}}
             at /home/joonas/src/neon/work/pageserver/src/tenant/tasks.rs:213:66
      pageserver::tenant::tasks::compaction_loop::{{closure}}
             at /home/joonas/src/neon/work/pageserver/src/tenant/tasks.rs:281:6
```

The line 763 maps the non-cancel variants (GetVectoredError and PageReconstructError).

File: libs/postgres_ffi/src/lib.rs [hunk 1]
```diff
@@ -143,8 +143,8 @@ pub use v14::xlog_utils::XLogFileName;
 
 pub use v14::bindings::DBState_DB_SHUTDOWNED;
 
-pub fn bkpimage_is_compressed(bimg_info: u8, version: u32) -> anyhow::Result<bool> {
-    dispatch_pgversion!(version, Ok(pgv::bindings::bkpimg_is_compressed(bimg_info)))
+pub fn bkpimage_is_compressed(bimg_info: u8, version: u32) -> bool {
+    dispatch_pgversion!(version, pgv::bindings::bkpimg_is_compressed(bimg_info))
 }
```
Comment (koivunej) [inline libs/postgres_ffi/src/lib.rs:148]: This always panicked, the anyhow wrapping was just making it seem like it might return an Err instead.

**Review by jcsp — APPROVED** [review]

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -663,34 +677,46 @@ pub(crate) enum GetVectoredError {
...
+    #[error("requested at invalid LSN: {0}")]
     InvalidLsn(Lsn),
 
-    #[error("Requested key not found: {0}")]
+    #[error("requested key not found: {0}")]
     MissingKey(MissingKeyError),
 
-    #[error(transparent)]
-    GetReadyAncestorError(GetReadyAncestorError),
+    #[error("ancestory walk")]
```
[the remark this answers is not in the document]
Comment (koivunej) [reply pageserver/src/tenant/timeline.rs:689]: oops!
```suggestion
    #[error("ancestry walk")]
```

**Pull request merged**